File: src/tui/input.ts
// Terminal key events -> tool key names (§6).
import type { KeyEvent } from "@opentui/core";
import { KEY } from "../core/tools/tool";
import { isPlaceable } from "../core/text";

const NAMED: Record<string, string> = {
  return: KEY.ENTER,
  enter: KEY.ENTER,
  linefeed: KEY.ENTER,
  backspace: KEY.BACKSPACE,
  delete: KEY.DELETE,
  up: KEY.UP,
  down: KEY.DOWN,
  left: KEY.LEFT,
  right: KEY.RIGHT,
};

/** The printable character a key produces, if any. */
export function printable(k: KeyEvent): string | null {
  if (k.ctrl || k.meta || k.super) return null;
  if (k.name === "space") return " ";
  if (k.sequence && isPlaceable(k.sequence)) return k.sequence;
  // Kitty-protocol events may carry an escape sequence; fall back to the name.
  if ([...k.name].length === 1) {
    const ch = k.shift ? k.name.toUpperCase() : k.name;
    if (isPlaceable(ch)) return ch;
  }
  return null;
}

/** Tool key for a terminal key: a printable character or an ASCIIFlow key name. */
export function toolKey(k: KeyEvent): string | null {
  return NAMED[k.name] ?? printable(k);
}

/** Alt+1..6 arrives as ESC+digit (meta) or as option on macOS with kitty keys. */
export function altDigit(k: KeyEvent): number | null {
  if (!(k.meta || k.option) || k.ctrl) return null;
  const n = Number(k.name);
  return Number.isInteger(n) && n >= 1 && n <= 6 ? n : null;
}

export const isCtrl = (k: KeyEvent, name: string) => k.ctrl && !k.meta && k.name === name;

File: src/tui/input.rs
//! Terminal key events -> tool keys.

use ratatui::crossterm::event::{KeyCode, KeyEvent, KeyModifiers};

use crate::core::text::is_placeable;
use crate::core::tools::tool::Key;

fn blocked(k: &KeyEvent) -> bool {
    k.modifiers
        .intersects(KeyModifiers::CONTROL | KeyModifiers::ALT | KeyModifiers::SUPER)
}

/// The printable character a key produces, if any.
pub fn printable(k: &KeyEvent) -> Option<char> {
    if blocked(k) {
        return None;
    }
    match k.code {
        KeyCode::Char(' ') => Some(' '),
        // Kitty-protocol events carry the shifted character already; legacy ones
        // report the unshifted one with SHIFT set.
        KeyCode::Char(c) if c.is_ascii_lowercase() && k.modifiers.contains(KeyModifiers::SHIFT) => {
            Some(c.to_ascii_uppercase())
        }
        KeyCode::Char(c) if is_placeable(c) => Some(c),
        _ => None,
    }
}

/// Tool key for a terminal key: a printable character or a named key.
pub fn tool_key(k: &KeyEvent) -> Option<Key> {
    match k.code {
        KeyCode::Enter => Some(Key::Enter),
        KeyCode::Backspace => Some(Key::Backspace),
        KeyCode::Delete => Some(Key::Delete),
        KeyCode::Up => Some(Key::Up),
        KeyCode::Down => Some(Key::Down),
        KeyCode::Left => Some(Key::Left),
        KeyCode::Right => Some(Key::Right),
        _ => printable(k).map(Key::Char),
    }
}

/// The `ctrl+<char>` binding a key carries, if any. Ctrl+Shift+Z reports as `z` too.
pub fn ctrl_char(k: &KeyEvent) -> Option<char> {
    if !k.modifiers.contains(KeyModifiers::CONTROL) || k.modifiers.contains(KeyModifiers::ALT) {
        return None;
    }
    match k.code {
        KeyCode::Char(c) if c.is_ascii_alphabetic() => Some(c.to_ascii_lowercase()),
        _ => None,
    }
}

pub fn is_ctrl(k: &KeyEvent, c: char) -> bool {
    ctrl_char(k) == Some(c)
}

/// Alt+1..6 arrives as ESC+digit (meta) or as option on macOS with kitty keys.
pub fn alt_digit(k: &KeyEvent) -> Option<usize> {
    if !k.modifiers.contains(KeyModifiers::ALT) || k.modifiers.contains(KeyModifiers::CONTROL) {
        return None;
    }
    match k.code {
        KeyCode::Char(c @ '1'..='6') => c.to_digit(10).map(|d| d as usize),
        _ => None,
    }
}

pub fn is_shift(k: &KeyEvent) -> bool {
    k.modifiers.contains(KeyModifiers::SHIFT)
        || matches!(k.code, KeyCode::Char(c) if c.is_ascii_uppercase())
}
